src/file_pipe_log/pipe_builder.rs

```diff
@@ -125,8 +169,12 @@ impl<B: FileBuilder> DualPipesBuilder<B> {
         Ok(())
     }
 
-    pub fn recover<S: ReplayMachine>(&mut self) -> Result<(S, S)> {
+    pub fn recover<S: ReplayMachine>(
+        &mut self,
+        machine_builder: &ReplayMachineBuilder<S>,
```

Comment:
Refer to https://github.com/tikv/raft-engine/pull/175#discussion_r772969268
```suggestion
    pub fn recover<F: ReplayMachineFactory>(
        &mut self,
        machine_builder: &F,
```